src/main/java/javaslang/collection/Vector.java

```diff
@@ -0,0 +1,1481 @@
+        public int hashCode() {
+            return hashCode.get();
+        }
+
+        @Override
+        public String toString() {
+            return map(String::valueOf).join(", ", "Vector(", ")");
+        }
+
+        @Override
+        public HashArrayMappedTrie<Integer, T> toHashArrayMappedTrie() {
+            return trie;
+        }
+
+        private Object writeReplace() {
+            return new SerializationProxy<>(this);
+        }
+
+        private void readObject(ObjectInputStream stream) throws InvalidObjectException {
+            throw new InvalidObjectException("Proxy required");
+        }
+
+        /**
+         * A serialization proxy.
+         * See {@link List.Cons.SerializationProxy}
+         */
+        private static final class SerializationProxy<T> implements Serializable {
+
+            private static final long serialVersionUID = 1L;
+
+            private transient Impl<T> vector;
+
+            SerializationProxy(Impl<T> vector) {
+                this.vector = vector;
+            }
+
+            private void writeObject(ObjectOutputStream s) throws IOException {
+                s.defaultWriteObject();
+                s.writeInt(vector.length());
+                for (T t: vector) {
+                    s.writeObject(t);
+                }
+            }
+
+            private void readObject(ObjectInputStream s) throws ClassNotFoundException, IOException {
+                s.defaultReadObject();
+                final int size = s.readInt();
+                if (size <= 0) {
+                    throw new InvalidObjectException("No elements");
+                }
+                HashArrayMappedTrie<Integer, T> temp = HashArrayMappedTrie.empty();
+                for (int i = 0; i < size; i++) {
+                    @SuppressWarnings("unchecked")
+                    final T element = (T) s.readObject();
+                    temp = temp.put(temp.size(), element);
+                }
+                vector = new Impl<>(temp);
+            }
+
+            private Object readResolve() {
+                return vector;
+            }
+        }
+    }
+
+    class Nil<T> implements Vector<T>, Serializable {
+        private static final long serialVersionUID = 1L;
+        private static final Vector<?> INSTANCE = new Nil<>();
+
+        private final HashArrayMappedTrie<Integer, T> trie = HashArrayMappedTrie.empty();
+
+        private Nil() {
+        }
+
+        @SuppressWarnings("unchecked")
+        public static <T> Nil<T> instance() {
+            return (Nil<T>) INSTANCE;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return o == this;
+        }
+
+        @Override
+        public int hashCode() {
+            return Traversable.hash(this);
+        }
+
+        @Override
+        public String toString() {
+            return "Vector()";
+        }
+
+        @Override
+        public HashArrayMappedTrie<Integer, T> toHashArrayMappedTrie() {
```

Comment:
I need <code>Vector.toHashArrayMappedTrie()</code> method to be able to implement 99% of code in interface body. This hack is not so beautiful but very effective